File: partiql-plan/src/main/java/org/partiql/plan/rex/RexCast.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.spi.types.PType;

import java.util.List;

/**
 * Logical cast expression abstract base class.
 */
public abstract class RexCast extends RexBase {

    /**
     * Creates a new cast instance.
     * @param operand operand rex (operand 0)
     * @param target target type
     * @return new cast instance
     */
    @NotNull
    public static RexCast create(@NotNull Rex operand, @NotNull PType target) {
        return new Impl(operand, target);
    }

    /**
     * Gets the operand rex.
     * @return operand rex (operand 0)
     */
    @NotNull
    public abstract Rex getOperand();

    /**
     * Gets the target type.
     * @return target type
     */
    @NotNull
    public abstract PType getTarget();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(getTarget());
    }

    @NotNull
    @Override
    protected List<Operand> operands() {
        Operand c0 = Operand.single(getOperand());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitCast(this, ctx);
    }

    private static class Impl extends RexCast {
        private final Rex operand;
        private final PType target;

        private Impl(@NotNull Rex operand, @NotNull PType target) {
            this.operand = operand;
            this.target = target;
        }

        @NotNull
        @Override
        public Rex getOperand() {
            return operand;
        }

        @NotNull
        @Override
        public PType getTarget() {
            return target;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexCast.kt
package org.partiql.plan.v1.rex

import org.partiql.types.PType

/**
 * Logical `CAST` operator — ex: CAST(<operand> AS <target>).
 */
public interface RexCast : Rex {

    public fun getOperand(): Rex

    public fun getTarget(): PType

    public override fun getOperands(): List<Rex> = listOf(getOperand())

    public override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitCast(this, ctx)

    /**
     * Default [RexCast] implementation meant for extension.
     */
    public abstract class Base(operand: Rex, target: PType) : RexCast {

        // DO NOT USE FINAL
        private var _operand = operand
        private var _target = target
        private var _operands: List<Rex>? = null

        public override fun getOperand(): Rex = _operand

        public override fun getTarget(): PType = _target

        public override fun getOperands(): List<Rex> {
            if (_operands == null) {
                _operands = listOf(_operand)
            }
            return _operands!!
        }

        public override fun getType(): PType = _target
        public override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is RexCast) return false
            if (_operand != other.getOperand()) return false
            if (_target != other.getTarget()) return false
            return true
        }

        public override fun hashCode(): Int {
            var result = 1
            result = 31 * result + _operand.hashCode()
            result = 31 * result + _target.hashCode()
            return result
        }
    }
}
